packages/compass-workspaces/src/stores/workspaces.ts

```diff
@@ -0,0 +1,519 @@
+    newTab: !!tabOptions?.newTab,
+  };
+};
+
+type SelectTabAction = { type: WorkspacesActions.SelectTab; atIndex: number };
+
+export const selectTab = (atIndex: number): SelectTabAction => {
+  return { type: WorkspacesActions.SelectTab, atIndex };
+};
+
+type MoveTabAction = {
+  type: WorkspacesActions.MoveTab;
+  fromIndex: number;
+  toIndex: number;
+};
+
+export const moveTab = (fromIndex: number, toIndex: number): MoveTabAction => {
+  return { type: WorkspacesActions.MoveTab, fromIndex, toIndex };
+};
+
+type SelectPreviousTabAction = { type: WorkspacesActions.SelectPreviousTab };
+
+export const selectPrevTab = (): SelectPreviousTabAction => {
+  return { type: WorkspacesActions.SelectPreviousTab };
+};
+
+type SelectNextTabAction = { type: WorkspacesActions.SelectNextTab };
+
+export const selectNextTab = (): SelectNextTabAction => {
+  return { type: WorkspacesActions.SelectNextTab };
+};
+
+type OpenTabFromCurrentActiveAction = {
+  type: WorkspacesActions.OpenTabFromCurrentActive;
+};
+
+export const openTabFromCurrent = (): OpenTabFromCurrentActiveAction => {
+  return { type: WorkspacesActions.OpenTabFromCurrentActive };
+};
+
+type CloseTabAction = { type: WorkspacesActions.CloseTab; atIndex: number };
+
+export const closeTab = (
+  atIndex: number
+): ThunkAction<void, WorkspacesState, void, CloseTabAction> => {
+  return (dispatch, getState) => {
+    const tab = getState().tabs[atIndex];
+    dispatch({ type: WorkspacesActions.CloseTab, atIndex });
+    cleanupLocalAppRegistryForTab(tab?.id);
+  };
+};
+
+type CollectionRemovedAction = {
+  type: WorkspacesActions.CollectionRemoved;
+  namespace: string;
+};
+
+export const collectionRemoved = (
+  namespace: string
+): WorkspacesThunkAction<void, CollectionRemovedAction> => {
+  return (dispatch, getState) => {
+    const tabsToRemove = getState().tabs.filter(
+      (tab) => tab.type === 'Collection' && tab.namespace === namespace
+    );
+    dispatch({
+      type: WorkspacesActions.CollectionRemoved,
+      namespace,
+    });
+    tabsToRemove.forEach((tab) => {
+      cleanupLocalAppRegistryForTab(tab.id);
+    });
+  };
+};
+
+type DatabaseRemovedAction = {
+  type: WorkspacesActions.DatabaseRemoved;
+  namespace: string;
+};
+
+export const databaseRemoved = (
+  namespace: string
+): WorkspacesThunkAction<void, DatabaseRemovedAction> => {
+  return (dispatch, getState) => {
+    const tabsToRemove = getState().tabs.filter((tab) => {
+      switch (tab.type) {
+        case 'Collections':
+          return tab.namespace === namespace;
+        case 'Collection':
+          return toNS(tab.namespace).database === namespace;
+        default:
+          return false;
+      }
+    });
+    dispatch({
+      type: WorkspacesActions.DatabaseRemoved,
+      namespace,
+    });
+    tabsToRemove.forEach((tab) => {
+      cleanupLocalAppRegistryForTab(tab.id);
+    });
+  };
+};
+
+// TODO: events are re-emitted when tab is changed for compatibility reasons,
```

Comment:
This is basically the biggest hack this PR has while sidebar and collections list plugins still relies on multiple events to get the current active workspaces state synced with its internal state. Removing the need for this will be the first thing I'll be addressing in the follow-up PRs